### PR TITLE
cloud-init: make cloud-init 'instance-id' persistent

### PR DESCRIPTION
**Pull request by acardace**
Before this patch the VMI uuid was used as the cloud-init `uuid`
(which is then used by cloud-init to build the `instance-id`),
this had the side-effect of letting the cloud-init `instance-id`
change on every VM reboot causing the 'on-first boot' cloud-init
configuration to be executed on every restart and not just once.

This for example would cause the VMI to regenerate SSH keys at
every restart which is not desirable (thus causing a ssh public key
mismatch when connnecting with a client).

This patch fixes this by using the VM UID as uuid which
makes the cloud-init 'instance-id' stable, in case the VMI is not
associated to a VM the VMI's UID will be used instead.

Signed-off-by: Antonio Cardace <acardace@redhat.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
make cloud-init 'instance-id' persistent across reboots
```

### CODE REVIEW

**Comment by lyarwood** [comment]
Wouldn't it be better to persist (or maybe initially generate and store) the VMI UUID in the underlying VM (if available) to ensure it remains stable in cloud-init?

Replacing a UUID with a random string doesn't strike me as the right approach here tbh.

**Comment by acardace** [comment]
> Wouldn't it be better to persist (or maybe initially generate and store) the VMI UUID in the underlying VM (if available) to ensure it remains stable in cloud-init?

I don't think that would be an option, the relationship goes the other way (VM owns the VMI), the VMI storing something in the VM doesn't sound great to me.

> 
> Replacing a UUID with a random string doesn't strike me as the right approach here tbh.
While I agree that there is surely a better fix as this doesn't exactly look like a UUID I don't think it's a random string, the combination VM_NAME.NAMESPACE is unique for the whole cluster.

I think the problem here is that the first implementation was done with the idea that the field `instance_id` would then become cloud-init's `instance-id` but actually is the `uuid` field that becomes the `instance-id` and here I'm merely "fixing it".

**Comment by lyarwood** [comment]
> > I think the problem here is that the first implementation was done with the idea that the field `instance_id` would then become cloud-init's `instance-id` but actually is the `uuid` field that becomes the `instance-id` and here I'm merely "fixing it".
> 
> I'd drop `UUID` entirely then and just leave cloud-init to default to the unique `instance-id`? Can you point to where this is happening btw?

https://github.com/kubevirt/kubevirt/blob/e8c9cacfde00a391d038d6161e17530fc1e92f4f/pkg/cloud-init/cloud-init.go#L577-L583

FWIW we already have a codepath that produces meta_data.json without the UUID.

**Comment by acardace** [comment]
> > > I think the problem here is that the first implementation was done with the idea that the field `instance_id` would then become cloud-init's `instance-id` but actually is the `uuid` field that becomes the `instance-id` and here I'm merely "fixing it".
> > 
> > 
> > I'd drop `UUID` entirely then and just leave cloud-init to default to the unique `instance-id`? Can you point to where this is happening btw?

It happens here
https://github.com/canonical/cloud-init/blob/1f0baf383cb1e3d93063bdc51780f00aa3f3da06/cloudinit/sources/helpers/openstack.py#L33
https://github.com/canonical/cloud-init/blob/1f0baf383cb1e3d93063bdc51780f00aa3f3da06/cloudinit/sources/helpers/openstack.py#L349

> 
> https://github.com/kubevirt/kubevirt/blob/e8c9cacfde00a391d038d6161e17530fc1e92f4f/pkg/cloud-init/cloud-init.go#L577-L583
> 
> FWIW we already have a codepath that produces meta_data.json without the UUID.

I've already tried that and the ssh key injection was not working but maybe I messed up something, let me try again as that would be indeed the best solution.

**Comment by acardace** [comment]
> FWIW we already have a codepath that produces meta_data.json without the UUID.

Unfortunately having just the 'istance-id' doesn't work as UUID is a required field, so actually that codepath is a bug that has to be fixed.

This is what happens https://github.com/canonical/cloud-init/blob/1f0baf383cb1e3d93063bdc51780f00aa3f3da06/cloudinit/sources/helpers/openstack.py#L350.

**Comment by acardace** [comment]
@lyarwood actually I can use the VM UUID which is stored in the `ownerReferences`! That's a real UUID and is stable.

**Comment by acardace** [comment]
/hold
I will add an e2e test as well, but this can be reviewed already.

File: pkg/cloud-init/cloud-init.go [hunk 1]
```diff
@@ -110,6 +110,15 @@ func IsValidCloudInitData(cloudInitData *CloudInitData) bool {
 	return cloudInitData != nil && cloudInitData.UserData != "" && (cloudInitData.NoCloudMetaData != nil || cloudInitData.ConfigDriveMetaData != nil)
 }
 
+func cloudInitUUIDFromVMI(vmi *v1.VirtualMachineInstance) string {
```
Comment (fabiand) [inline pkg/cloud-init/cloud-init.go:113]: What about using `VMI.spec.firmware.uuid`
IIRC then this value is derived from the name by default, thus it does not change between restarts, and should even stay stable if VMI is managed bby a VM, yet it will be different if the VM is cloned.

To me using this field would make the solution more self contained.

Comment (lyarwood) [reply pkg/cloud-init/cloud-init.go:113]: While it's a valid suggestion to use `VMI.spec.firmware.uuid` wouldn't it be slightly confusing for users if it was provided to cloud-init as the instance UUID? I'd much rather see the code below used to return either the VirtualMachine UUID or VirtualMachineInstance UUID as not to confuse users who might end up encountering it used as if it was the instance UUID.

Comment (acardace) [reply pkg/cloud-init/cloud-init.go:113]: @lyarwood why do you think it's confusing? I actually think it's better to use that because as Fabian said it's self contained.

What we need for the instance-id is just a stable UUID and VMI.spec.firmware.uuid is derived from the VM/VMI name and is stable.

**Comment by acardace** [comment]
/unhold
e2e test added

File: pkg/cloud-init/cloud-init.go (hunk 1 above)
Comment (fabiand) [reply pkg/cloud-init/cloud-init.go:115]: Sorry, I missed Lee's comment - Lee, in KubeVirt's terms, the VMI firmware uuid is derived from the VMI name (by default), thus I'm seeing the firmware uuid to really be the instance id - because they are usually coupled. WDYT?

File: pkg/virt-controller/watch/snapshot/restore.go [hunk 2]
```diff
@@ -523,6 +537,8 @@ func (t *vmRestoreTarget) Reconcile() (bool, error) {
 			Spec:   *snapshotVM.Spec.DeepCopy(),
 			Status: kubevirtv1.VirtualMachineStatus{},
 		}
+
+		stripIdentityInfo(newVM)
```
Comment (iholder-redhat) [inline pkg/virt-controller/watch/snapshot/restore.go:541]: I think this is problematic. Let me give you some context first:

In order to [implement the cloning API](https://github.com/kubevirt/kubevirt/pull/7336) I first needed to [adapt](https://github.com/kubevirt/kubevirt/pull/7318) restore/snapshot controllers. The two are supposed to work in the following way:
* The snapshot / restore controllers can restore a snapshot into a brand new VM (if it doesn't exist).
* When this is being done, the restore creator can provide a set of patches (through the restore's spec).
  * The restore mechanism does not know how to strip a VM's identity. All it can do is get patches through spec and apply them on the target VM (offline, before it's created). This keeps a clear separation of concerns between the restore and clone APIs.
* The clone API is responsible to provide the right patches so the identity would be stripped properly.

What's needed to be done here is to add the proper patch into restore's functional tests. This way no identity stripping logic resides in the restore controller.

But maybe this can be done in a follow-up PR.

File: pkg/cloud-init/cloud-init.go [hunk 3]
```diff
@@ -110,6 +112,13 @@ func IsValidCloudInitData(cloudInitData *CloudInitData) bool {
...
+	if vmi.Spec.Domain.Firmware == nil {
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply pkg/cloud-init/cloud-init.go:116]: I agree. :+1:

File: pkg/cloud-init/cloud-init.go [hunk 4]
```diff
@@ -537,14 +547,14 @@ func GenerateEmptyIso(vmiName string, namespace string, data *CloudInitData, siz
 	return nil
 }
 
-func GenerateLocalData(vmiName string, namespace string, instanceType string, data *CloudInitData) error {
-	precond.MustNotBeEmpty(vmiName)
+func GenerateLocalData(vmi *v1.VirtualMachineInstance, instanceType string, data *CloudInitData) error {
```
Comment (iholder-redhat) [inline pkg/cloud-init/cloud-init.go:550]: nit: no real need to change the parameters here

Comment (iholder-redhat) [reply pkg/cloud-init/cloud-init.go:550]: EDIT: ah, it's for `cloudInitUUIDFromVMI`. Sorry

File: tests/console/console.go [hunk 5]
```diff
@@ -129,6 +130,36 @@ func RunCommand(vmi *v1.VirtualMachineInstance, command string, timeout time.Dur
 	return nil
 }
 
+// RunCommandAndStoreOutput runs the command line from `command` connecting to an already logged in console in vmi.
+// The output of `command` is returned as a string
+func RunCommandAndStoreOutput(vmi *v1.VirtualMachineInstance, command string, timeout time.Duration) (string, error) {
```
Comment (EdDev) [inline tests/console/console.go:135]: This seems to be very similar to the existing `SafeExpectBatchWithResponse` function.
Is there a specific reason to have this additional one with all this login duplicated?

I suspect this introduces a potential flakiness due to the missing checks (that we do have a console, that the command did reach the console, etc).
